src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/UnixPkcs12Reader.cs

```diff
@@ -0,0 +1,670 @@
+                    // compare SubjectPublicKeyInfo values
+                    if (matchingKeyIdx == -1 && keyBagIdx > 0)
+                    {
+                        ICertificatePalCore cert = _certs[certBagIdx].Cert;
+                        string algorithm = cert.KeyAlgorithm;
+                        byte[] keyParams = cert.KeyAlgorithmParameters;
+                        byte[] keyValue = cert.PublicKeyValue;
+
+                        for (int i = 0; i < keyBagIdx; i++)
+                        {
+                            if (PublicKeyMatches(algorithm, keyParams, keyValue, ref publicKeyInfos[i]))
+                            {
+                                matchingKeyIdx = i;
+                                break;
+                            }
+                        }
+                    }
+
+                    if (matchingKeyIdx != -1)
+                    {
+                        if (keys[matchingKeyIdx] == null)
+                        {
+                            throw new CryptographicException(SR.Cryptography_Pfx_BadKeyReference);
+                        }
+
+                        _certs[certBagIdx].Key = keys[matchingKeyIdx];
+                        keys[matchingKeyIdx] = null;
+                    }
+                }
+
+                _certCount = certsCount;
+            }
+            catch
+            {
+                if (_certs != null)
+                {
+                    foreach (CertAndKey certAndKey in _certs)
+                    {
+                        certAndKey.Dispose();
+                    }
+                }
+
+                throw;
+            }
+            finally
+            {
+                if (keys != null)
+                {
+                    foreach (AsymmetricAlgorithm key in keys)
+                    {
+                        key?.Dispose();
+                    }
+
+                    ArrayPool<AsymmetricAlgorithm>.Shared.Return(keys);
+                }
+
+                if (publicKeyInfos != null)
+                {
+                    ArrayPool<SubjectPublicKeyInfoAsn>.Shared.Return(publicKeyInfos, clearArray: true);
+                }
+
+                ArrayPool<CertBagAsn>.Shared.Return(certBags, clearArray: true);
+                ArrayPool<AttributeAsn[]>.Shared.Return(certBagAttrs, clearArray: true);
+                ArrayPool<SafeBagAsn>.Shared.Return(keyBags, clearArray: true);
+            }
+        }
+
+        private static bool PublicKeyMatches(
+            string algorithm,
+            byte[] keyParams,
+            byte[] keyValue,
+            ref SubjectPublicKeyInfoAsn publicKeyInfo)
+        {
+            if (!publicKeyInfo.SubjectPublicKey.Span.SequenceEqual(keyValue))
+            {
+                return false;
+            }
+
+            switch (algorithm)
+            {
+                case Oids.Rsa:
+                case Oids.RsaPss:
+                    switch (publicKeyInfo.Algorithm.Algorithm.Value)
```

Comment:
should this be:

```csharp
if (publicKeyInfo.Algorithm.Algorithm.Value != algorithm)
    return false;
```

Reply:
(same comment for EcPublicKey/EcDiffieHellman)

Reply:
although looking further this might be intentional

Reply:
The most common pairing will be (Rsa, Rsa) then (RsaPss, Rsa), then maybe (RsaPss, RsaPss) and I-can't-really-imagine-it-but-it's-still-sensible (Rsa, RsaPss).  (All tuples were (certAlg, keyAlg)).

Similarly for ECDH/ECDSA.